### PR TITLE
Refactored state_filter functionality for tabbedview listings.

### PR DESCRIPTION
**Pull request by phgross**
The current implementation for the `Active` `All` filter is circuitous and very limited to the current use case. Because we'll extend the dossier listing filters with an additional one, I've had to rework the current implementation.

Therefore I've created new objects `Filter` and `FilterList` which should make the code easier to understand and read and allow additional and more complicated filters. 

The changes are only refactoring, the functionality itself is not changed.

@deiferni @lukasgraf please review ...

### CODE REVIEW

File: opengever/tabbedview/browser/generic_with_filters.pt [hunk 1]
```diff
@@ -46,23 +46,21 @@
...
-            tal:attributes="id state_filter_name" class="state_filters">
+      <span tal:condition="view/filterlist_available"
+            tal:attributes="id filterlist_name" class="state_filters">
         <span i18n:translate="" >State</span>
 
-        <a id="filter_all"
-           tal:attributes="class python: all_filter and 'active'"
+        <a tal:repeat="filter python: view.filterlist.filters()"
+           tal:attributes="id filter/id;
+                           class python: (filter.id == filter_id or (filter_id is None and filter.default)) and 'active'"
```
Comment (deiferni) [inline opengever/tabbedview/browser/generic_with_filters.pt]: Too much logic. Please move this to the View, or you can ask a `Filter` about its class?

File: opengever/tabbedview/browser/selection_with_filters.pt [hunk 2]
```diff
@@ -12,16 +12,21 @@
...
+          tal:attributes="id filterlist_name" class="state_filters">
 
       <span> | </span>
 
       <span i18n:translate="" i18n:domain="opengever.tabbedview">State</span>
 
-      <a id="filter_all" tal:attributes="class python: all_filter and 'active'" href="javascript:void(0);" i18n:domain="opengever.tabbedview" i18n:translate="">All</a>
+      <a tal:repeat="filter python: view.filterlist.filters()"
+         tal:attributes="id filter/id;
+                         class python: (filter.id == filter_id or (filter_id is None and filter.default)) and 'active'"
```
Comment (deiferni) [inline opengever/tabbedview/browser/selection_with_filters.pt]: Same here.

File: opengever/tabbedview/filters.py [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+    def update_query(self, query):
+        return query
+
+
+class CatalogQueryFilter(Filter):
+    """A Filter class which extends the existing portal_catalog query with the
+    given extension `query_extension`.
+    """
+
+    def __init__(self, id, label, default=False, query_extension={}):
```
Comment (lukasgraf) [inline opengever/tabbedview/filters.py]: [Mutable default argument](http://effbot.org/zone/default-values.htm) - I'd rather write this as
```python
def __init__(self, id, label, default=False, query_extension=None):
    if query_extension is None:
        query_extension = {}
```

File: opengever/tabbedview/filters.py [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+
+
+class FilterList(object):
+    """A list of tabbedview listing Filter objects.
+    """
+
+    def __init__(self, *args):
+        self._filters = OrderedDict()
+        self.default_filter = None
+        for filter in args:
```
Comment (lukasgraf) [inline opengever/tabbedview/filters.py]: `filter` is shadowing the a builtin `filter()` function. -> `filter_` or `flt`?

File: opengever/tabbedview/filters.py [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+            self._filters[filter.id] = filter
+            if filter.default:
+                if self.default_filter:
+                    raise ValueError(
+                        'Only one Filter marked as default possible.')
```
Comment (deiferni) [inline opengever/tabbedview/filters.py]: Only one ~~F~~filter marked as default **is** possible.

File: opengever/tabbedview/browser/tabs.py [hunk 6]
```diff
@@ -226,26 +239,24 @@ def build_query(self):
...
 
-    def extend_query_with_statefilter(self, query):
+    def extend_query_with_filter(self, query):
         """Extends the given query with a filter,
         which show just objects in the open state."""
 
-        query['review_state'] = self.config.open_states
-        return query
+        filter_id = self.request.get(self.config.filterlist_name, None)
+        if filter_id:
```
Comment (lukasgraf) [inline opengever/tabbedview/browser/tabs.py]: Maybe rewrite like this?
```python
filter_id = self.request.get(self.config.state_filter_name, None)
filter_ = self.config.filterlist.default_filter

if filter_id:
    filter_ = self.config.filterlist.get(filter_id)

return filter_.update_query(query)
```

File: opengever/tabbedview/filters.py [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+
+                self.default_filter = filter
+
+    def get(self, filter_id):
```
Comment (lukasgraf) [inline opengever/tabbedview/filters.py]: This could potentially be implemented as `__getitem__`...

**Comment by lukasgraf** [comment]
Commit msg in 3. commit: `Update Naming from StateFilter... to FitlerList` -> `FilterList`

**Comment by lukasgraf** [comment]
Other than the mentioned points: :+1: :green_heart: :cake:

**Comment by phgross** [comment]
@lukasgraf @deiferni PR Updated and extend please have another look.

File: opengever/tabbedview/browser/tabs.py [hunk 8]
```diff
@@ -226,26 +239,22 @@ def build_query(self):
...
 
-    def extend_query_with_statefilter(self, query):
+    def extend_query_with_filter(self, query):
         """Extends the given query with a filter,
         which show just objects in the open state."""
 
-        query['review_state'] = self.config.open_states
-        return query
+        selected_filter_id = self.request.get(
+            self.config.filterlist_name, None)
```
Comment (lukasgraf) [inline opengever/tabbedview/browser/tabs.py]: Default of `None` is unnecessary - `.get()` always defaults to `None`.

File: opengever/tabbedview/browser/tasklisting.py [hunk 9]
```diff
@@ -154,21 +162,18 @@ def build_query(self):
...
-    def extend_query_with_statefilter(self, query):
-        """When a state filter is active,
-        we add a filter which selects just the pending tasks"""
-
-        return query.in_pending_state()
+    def extend_query_with_filter(self, query):
+        """When the filterlist is active, we update the query with
+        the current filter."""
+        selected_filter_id = self.request.get(
+            self.config.filterlist_name, None)
```
Comment (lukasgraf) [inline opengever/tabbedview/browser/tasklisting.py]: Same here, drop `None`

**Comment by lukasgraf** [comment]
:+1:

**Pull request merged**